core/testsuite/src/test/hibernate/com/blazebit/persistence/testsuite/DereferenceManyToOneAssociationInJoinedInheritanceHierarchyTest.java

```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2014 - 2019 Blazebit.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.blazebit.persistence.testsuite;
+
+import com.blazebit.persistence.CriteriaBuilder;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate42;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate43;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate50;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate51;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate52;
+import com.blazebit.persistence.testsuite.base.jpa.category.NoHibernate53;
+import org.junit.Assume;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+
+import javax.persistence.Column;
+import javax.persistence.Entity;
+import javax.persistence.GeneratedValue;
+import javax.persistence.GenerationType;
+import javax.persistence.Id;
+import javax.persistence.Inheritance;
+import javax.persistence.InheritanceType;
+import javax.persistence.ManyToOne;
+import javax.persistence.Tuple;
+
+import java.util.Properties;
+
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * @author Jan-Willem Gmelig Meyling
+ * @since 1.4.0
+ */
+public class DereferenceManyToOneAssociationInJoinedInheritanceHierarchyTest extends AbstractCoreTest {
```

Comment:
IMO the tests need to make use of data to verify the desired behavior. The assertions should be the same for all the tests.

A/B | id | parent_id
A    |  1  | null
A    |  2  | null
A    |  3  | 1
B    |  4  | null
B    |  5  | 3
B    |  6  | 4  <-- wrong data on purpose
B    |  7  | 10  <-- wrong data on purpose

Results

C1 | C2
4   |  null
5   |  3
6   |  null
7   |  null

Reply:
Ive verified the generated SQL for all versions of Hibernate, but I can add some additional assertions too.

Reply:
I don't think the tests for the "wrong data" entries are sensible. The generic bound only affects setters and getters, not the actual target type. So in the schema / for hibernate its perfectly valid for a B to reference another B. Or do you just want to see that this B is ommitted, as we're actively joining against A?

Reply:
Correct, B(6) and B(7) are IMO edge cases that might produce wrong results when implemented in a wrong way